File: aiqcomponents/src/main/java/com/aiqfome/aiqcomponents/selector/SelectorController.java
package com.aiqfome.aiqcomponents.selector;

import android.os.Handler;
import android.view.View;

import androidx.fragment.app.FragmentManager;

import com.aiqfome.aiqcomponents.adapters.BaseItem;
import com.aiqfome.aiqcomponents.adapters.CommonAdapter;
import com.aiqfome.aiqcomponents.adapters.CommonSearchableAdapter;
import com.aiqfome.aiqcomponents.sheets.ListBottomSheet;
import com.aiqfome.aiqcomponents.sheets.SearchableListBottomSheet;

import java.util.List;

public abstract class SelectorController<Type> {

    private Selector selector;

    private FragmentManager fragmentManager;
    private String title;
    private List<BaseItem<Type>> itemList;

    private boolean shouldDismissOnSelect = true;
    private boolean isFirstItemPreSelected = false;
    private boolean displaySelectedSubtitle = false;

    private Type selectedItem;

    private CommonAdapter adapter;
    private ListBottomSheet listBottomSheet;

    private CommonSearchableAdapter searchableAdapter;
    private SearchableListBottomSheet searchableListBottomSheet;

    private boolean isSearchable = false;

    public SelectorController(FragmentManager fragmentManager,
                              String title,
                              List<BaseItem<Type>> itemList) {

        this.fragmentManager = fragmentManager;
        this.title = title;
        this.itemList = itemList;

        setup();
    }

    public SelectorController(FragmentManager fragmentManager,
                              String title,
                              List<BaseItem<Type>> itemList,
                              boolean isSearchable) {

        this.fragmentManager = fragmentManager;
        this.title = title;
        this.itemList = itemList;
        this.isSearchable = isSearchable;

        setup();
    }

    public SelectorController(FragmentManager fragmentManager,
                              String title,
                              List<BaseItem<Type>> itemList,
                              boolean displaySelectedSubtitle,
                              boolean isSearchable) {

        this.fragmentManager = fragmentManager;
        this.title = title;
        this.itemList = itemList;
        this.displaySelectedSubtitle = displaySelectedSubtitle;
        this.isSearchable = isSearchable;

        setup();
    }

    public SelectorController(FragmentManager fragmentManager,
                              String title,
                              List<BaseItem<Type>> itemList,
                              boolean shouldDismissOnSelect,
                              boolean isFirstItemPreSelected,
                              boolean isSearchable) {

        this.fragmentManager = fragmentManager;
        this.title = title;
        this.itemList = itemList;
        this.shouldDismissOnSelect = shouldDismissOnSelect;
        this.isFirstItemPreSelected = isFirstItemPreSelected;
        this.isSearchable = isSearchable;

        setup();
    }

    public SelectorController(FragmentManager fragmentManager,
                              String title,
                              List<BaseItem<Type>> itemList,
                              boolean shouldDismissOnSelect,
                              boolean isFirstItemPreSelected,
                              boolean displaySelectedSubtitle,
                              boolean isSearchable) {

        this.fragmentManager = fragmentManager;
        this.title = title;
        this.itemList = itemList;
        this.shouldDismissOnSelect = shouldDismissOnSelect;
        this.isFirstItemPreSelected = isFirstItemPreSelected;
        this.displaySelectedSubtitle = displaySelectedSubtitle;
        this.isSearchable = isSearchable;

        setup();
    }

    private void setup() {

        if (isSearchable) {

            searchableAdapter = new CommonSearchableAdapter<Type>() {
                @Override
                public View.OnClickListener itemClickListener(BaseItem<Type> item) {
                    return v -> {
                        selectItem(item);
                        if (shouldDismissOnSelect) dismiss();
                    };
                }
            };

            searchableAdapter.setBaseItemList(itemList);

            searchableListBottomSheet = new SearchableListBottomSheet(title, searchableAdapter);

        } else {

            adapter = new CommonAdapter<Type>() {
                @Override
                public View.OnClickListener itemClickListener(BaseItem<Type> item) {
                    return v -> {
                        selectItem(item);
                        if (shouldDismissOnSelect) dismiss();
                    };
                }
            };

            adapter.setBaseItemList(itemList);

            listBottomSheet = new ListBottomSheet(title, adapter);
        }

    }

    private void selectItem(BaseItem<Type> item) {
        if (selector != null)
            selector.setSelectedItem(displaySelectedSubtitle ? item.getSubTitle() : item.getTitle());

        selectedItem = item.getObject();

        onItemSelected(item.getObject());
    }

    public abstract void onItemSelected(Type object);

    void setSelector(Selector selector) {
        this.selector = selector;
        if (isFirstItemPreSelected && !itemList.isEmpty()) selectItem(itemList.get(0));
    }

    void show() {
        if (isSearchable) {
            if (!searchableListBottomSheet.isAdded())
                searchableListBottomSheet.show(fragmentManager,
                        title + "searchableListBottomSheet");
        } else {
            if (!listBottomSheet.isAdded())
                listBottomSheet.show(fragmentManager, title + " listBottomSheet");
        }
    }

    public void dismiss() {
        if (isSearchable)
            new Handler().postDelayed(() -> searchableListBottomSheet.dismissAllowingStateLoss(),
                    200);
        else
            new Handler().postDelayed(() -> listBottomSheet.dismissAllowingStateLoss(),
                    200);
    }

    public List<BaseItem<Type>> getItemList() {
        return itemList;
    }

    public Type getSelectedItem() {
        return selectedItem;
    }

}

File: aiqcomponents/src/main/java/com/aiqfome/aiqcomponents/selector/SelectorController.kt
package com.aiqfome.aiqcomponents.selector

import androidx.fragment.app.FragmentManager
import com.aiqfome.aiqcomponents.adapters.TextListAdapter
import com.aiqfome.aiqcomponents.adapters.TextListSearchableAdapter
import com.aiqfome.aiqcomponents.adapters.OnItemClickListener
import com.aiqfome.aiqcomponents.adapters.model.Item
import com.aiqfome.aiqcomponents.controller.BaseController
import com.aiqfome.aiqcomponents.sheets.ListBottomSheet
import com.aiqfome.aiqcomponents.sheets.SearchableListBottomSheet

abstract class SelectorController<T> @JvmOverloads constructor(
    private val fragmentManager: FragmentManager,
    private val title: String,
    private val list: List<Item<T>>,
    private val isSearchable: Boolean = false,
    private val shouldDismissOnSelect: Boolean = true,
    private val isFirstItemPreSelected: Boolean = false,
    private val displaySelectedSubtitle: Boolean = false,
) : BaseController<T>() {
    private var selector: Selector? = null

    init {
        setup()
    }

    final override fun setup() {
        if (isSearchable) {
            adapter = TextListSearchableAdapter()
            bottomSheet = (adapter as? TextListSearchableAdapter)?.let {
                SearchableListBottomSheet(title, it)
            }!!
        } else {
            adapter = TextListAdapter()
            bottomSheet = (adapter as? TextListAdapter)?.let {
                ListBottomSheet(title, it)
            }!!
        }

        adapter.onItemClickListener = OnItemClickListener { item ->
            selectItem(item)
            if (shouldDismissOnSelect) dismiss()
        }

        adapter.submitList(list)
    }

    override fun selectItem(item: Item<T>) {
        selector?.setSelectedItem(
            if (displaySelectedSubtitle) {
                item.subTitle
            } else item.title
        )

        onItemSelected(item.item)
    }

    fun setSelector(selector: Selector?) {
        this.selector = selector

        if (isFirstItemPreSelected && list.isNotEmpty()) {
            selectItem(list[FIRST_ITEM])
        }
    }

    override fun show() {
        if (!bottomSheet.isAdded) {
            val tag = title + if (isSearchable) {
                SEARCHABLE_LIST_BOTTOM_SHEET
            } else LIST_BOTTOM_SHEET
            bottomSheet.show(
                fragmentManager,
                tag
            )
        }
    }
}